### ISSUE TITLE
Possible memory leak, when deleting nodes in graphs repeatedly

### ISSUE THREAD

**Issue by Luxxii** [igraph/python-igraph#503]
Hi everyone,

i discovered a strange behavior, while using the newest igraph-version on PyPI. It looks like that if we repeatedly delete nodes in the same process for new graphs, the process will consume more and more RAM.

**Describe the bug**
To clarify, lets assume we use a process in Python, which generates graphs via igraph. The process generates thousands of graphs. It uses the `delete_vertices` method on each of those graphs. The problem now is, the more graphs this process generates, the more RAM the process will need.  It seems like the method `delete_vertices` somehow does not release the used memory (potential memory leak?).

**To reproduce**
It can be easily reproduced. I have written three small scripts, showcasing this strange behaviour. The scripts can be found [here](https://github.com/Luxxii/igraph_memory_leak).

I used `memory_profiler` to track used RAM-Comsumption. (If you are unfamiliar with `memory_profiler`, simply install it via `pip install memory_profiler`and run `mprof plot <mprofiler_file.dat>`)

Here is the code of the initial first test: 

`test.py` 
```python
import igraph 
import random

how_many_graphs = 20000  # How many graphs to generate
max_nodes = 1000
max_edges = 10000
delete_nodes_perc = 0.1  # Delete 10% of nodes!

for i in range(how_many_graphs):
    num_nodes = random.randint(1, max_nodes)
    num_edges = random.randint(1, max_edges)

    # Generate random graph
    graph = igraph.Graph(directed=True)
    graph.add_vertices(num_nodes)
    edges = [
        random.choices(range(num_nodes), k=2) 
        for _ in range(num_edges)
    ]
    graph.add_edges(edges)

    # !!! This part is problematic. In long running processes, more and more RAM is consumed.
    # delete nodes
    del_list = random.choices(range(num_nodes), k= int(num_nodes*delete_nodes_perc))
    graph.delete_vertices(del_list)
```
This script generates random graphs (with up to `max_nodes` nodes and `max_edges` edges) and then deletes randomly nodes (up to a specific percentage). The last bit of the script ( `graph.delete_vertices(del_list)`) deletes vertices which results into not releasing memory. This can be tracked over a certain period of time. The more graphs are generated, the more RAM is consumed. You can check this yourself by increasing `how_many_graphs`.

Here is a plot of the memory usage of this script, where it can be clearly seen, that the memory consumption increases linearly:

![test py](https://user-images.githubusercontent.com/18737631/150695113-ccd97a5f-b34b-4df1-91dd-7e8e57b6e139.png)

If deleting the graph explicitly via `del`, the memory consumption still increases linearly ([here](https://github.com/Luxxii/igraph_memory_leak/blob/main/test_del_graph_afterwards.py.png))

---

I also tested it with the `delete_edges`-method, the memory usage was in that case constant  ([here](https://github.com/Luxxii/igraph_memory_leak/blob/main/test_del_edges.py.png)).

---

**Version information**
Here is the `pip list`:
```text
Package         Version
--------------- -------
cycler          0.11.0
fonttools       4.28.5
igraph          0.9.9
kiwisolver      1.3.2
matplotlib      3.5.1
memory-profiler 0.60.0
numpy           1.22.1
packaging       21.3
Pillow          9.0.0
pip             21.3.1
psutil          5.9.0
pyparsing       3.0.7
python-dateutil 2.8.2
setuptools      59.6.0
six             1.16.0
texttable       1.6.4
wheel           0.37.0
(Python 3.10)
```

I can confirm this behaviour on `igraph 0.9.9` and on multiple different machines (all linux-based). I am not sure if other versions also have this behavior. I know that older versions of igraph did not had increasing RAM-consumption, but this was with the prior PyPI releases which were on `python-igraph`, which i could not remember. If wanted, i could try to check this over some igraph-versions.

**My Guess**
I guess that it may have something to do with deallocating memory on C, where when deleting an node, less memory is deallocated for nodes afterwards "forgetting" the deleted node, but i am no expert in C. I would appreciate if someone could check this. Maybe i am also doing something wrong.

**Comment by ntamas** [comment]
Confirmed on macOS with the development version. Let me check the code corresponding to the vertex deletion to see if I can spot anything obvious before digging into this further.

**Comment by ntamas** [comment]
I have fixed a leak in `igraphmodule_PyObject_to_vs_t()` that is used to convert a Python list of vertex IDs to an `igraph_vs_t` object in C. This reduces the memory consumption significantly according to my measurements, but I still see a slight upwards trending slope on the memory consumption curve when I run the test for longer (i.e. 50000 graphs or more). I'm not sure whether that's a genuine leak or an artifact in Python's memory management and this memory would eventually be returned to the OS if certain conditions are met. (I suspect this because I see no similar slope on your plots when using `delete_edges()`, but I see it on my machine with `delete_edges()`). I'd appreciate it if you could test the head of the `master` branch on your machine as well to see if it resolves the problem.

This is how the memory consumption plot looks like after the fix:

![Figure_1](https://user-images.githubusercontent.com/195637/150803505-ec670d4b-2200-4962-811d-4ccc31041336.png)

**Comment by ntamas** [comment]
Thanks for the feedback; the fix will be released in 0.9.10 or 0.10.0, whichever comes first.
